xarray/core/alignment.py

```diff
@@ -358,7 +366,15 @@ def var_indexers(var, indexers):
             args = (var.attrs, var.encoding)
         else:
             args = ()
-        reindexed[dim] = IndexVariable((dim,), indexers[dim], *args)
+
+        idx_var = IndexVariable((dim,), indexers[dim], *args)
+
+        # GH1434
+        # ensures that dtype of numpy structured arrays is preserved
+        if len(args) and is_composite_dtype(var.dtype) \
+                and idx_var.dtype != var.dtype:
+            idx_var.data = idx_var.data.astype(var.dtype)
```

Comment:
To fix this, we need a dtype safe version of `pandas.Index.__or__`, which is used for joining together indexes: https://github.com/maciekswat/xarray/blob/7cec75c6252c34aee46f3e7bb8c170c169429f37/xarray/core/alignment.py#L20

The cleanest way I see to handle this is to make the logic inside `align` always use some sort of xarray wrapper around `pandas.Index` similar to (or building on) [`PandasIndexAdapter`](https://github.com/maciekswat/xarray/blob/7cec75c6252c34aee46f3e7bb8c170c169429f37/xarray/core/indexing.py#L494) that propagates dtype properly in these operations. I don't want to add dtype specific logic in `align`/`reindex_variables` that fixes things up after the fact, because the result is much less maintainable.